core/src/main/java/org/apache/iceberg/BaseIncrementalChangelogScan.java

```diff
@@ -133,51 +131,149 @@ private static Map<Long, Integer> computeSnapshotOrdinals(Deque<Snapshot> snapsh
+                DeleteFile[] addedDeleteFiles = filterAdded(context.deletes().forEntry(entry));
+
+                switch (entry.status()) {
+                  case ADDED:
+                    if (entrySnapshotId == snapshotId) {
+                      return new BaseAddedRowsScanTask(
+                          changeOrdinal,
+                          snapshotId,
+                          dataFile,
+                          addedDeleteFiles,
+                          context.schemaAsString(),
+                          context.specAsString(),
+                          context.residuals());
+                    } else {
+                      // the data file is added before the snapshot we're processing
+                      if (addedDeleteFiles.length == 0) {
+                        return DummyChangelogScanTask.INSTANCE;
+                      } else {
+                        return new BaseDeletedRowsScanTask(
+                            changeOrdinal,
+                            snapshotId,
+                            dataFile,
+                            addedDeleteFiles,
+                            filterExisting(context.deletes().forEntry(entry)),
+                            context.schemaAsString(),
+                            context.specAsString(),
+                            context.residuals());
+                      }
+                    }
+
+                  case DELETED:
+                    if (entrySnapshotId == snapshotId) {
+                      return new BaseDeletedDataFileScanTask(
+                          changeOrdinal,
+                          snapshotId,
+                          dataFile,
+                          filterExisting(context.deletes().forEntry(entry)),
+                          context.schemaAsString(),
+                          context.specAsString(),
+                          context.residuals());
+                    } else {
+                      return DummyChangelogScanTask.INSTANCE;
+                    }
+
+                  case EXISTING:
+                    if (addedDeleteFiles.length == 0) {
+                      return DummyChangelogScanTask.INSTANCE;
+                    } else {
+                      return new BaseDeletedRowsScanTask(
+                          changeOrdinal,
+                          snapshotId,
+                          dataFile,
+                          addedDeleteFiles,
+                          filterExisting(context.deletes().forEntry(entry)),
+                          context.schemaAsString(),
+                          context.specAsString(),
+                          context.residuals());
+                    }
+
+                  default:
+                    throw new IllegalArgumentException(
+                        "Unexpected entry status: " + entry.status());
+                }
+              });
+      return CloseableIterable.filter(tasks, task -> !(task instanceof DummyChangelogScanTask));
```

Comment:
Would `task != DummyChangelogScanTask.INSTANCE` work?

Reply:
Yes, that works too. `instanceof` was just how I thought of it logically, and as I understand it, `instanceof` is highly optimized so performance is probably no worse.

Reply:
In my experience using `instanceof` is a sign of bad code when we don't use inheritance/object oriented principles correctly. I understand it that this is not the case here, but when reading the code, every new reader must learn/understand it, which is a small extra effort from every future developer. Also I would be surprised if the performance would be better than default implementation of the `equals` 😄

Reply:
Thanks for changing!